### PR TITLE
GDScript: Rework type check

### LINKED ISSUE #54312: GDScript 2.0: Type testing (`is`) not working with typed arrays
### Godot version

v4.0.dev.custom_build.3bebbcacd

### System information

manjaro linux

### Issue description

```
print([1,2] is Array[int])
```
yields parse error:
```
  built-in:5 - Parse Error: Builtin type cannot be used as a name on its own.
  built-in:5 - Parse Error: Identifier "int" not declared in the current scope.
  modules/gdscript/gdscript.cpp:855 - Method/function failed. Returning: ERR_PARSE_ERROR
```

### Steps to reproduce

Run the above

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by vonagam**
Handle type check - `a is A` - as its own thing, not as after-thought in logic of binary operator. A type is now parsed as a proper type specifier, not an expression. 

Add support for typed arrays.
```gdscript
a is Array[int] # now possible
```

Add support for checking with `Script` type, not only `GDScript`.

Fix handling of constants. Look at a type of actual reduced value instead of a declared type. Also fix constant null handling.
```gdscript
class A: pass

const three: Variant = 3
print(three is A) # before: true, now: false

print(null is A) # before: true, now: false
```

Since right side now needs to be compile time known type to support previously available functionality, where a type was intended to be variable, add `is_instance_of` utility function to GDScript.
```gdscript
is_instance_of(a, TYPE_INT)
is_instance_of(a, RefCounted)
is_instance_of(a, A)
```

In `GDScriptAnalyzer::resolve_datatype` do not set `builtin_type` to `Variant::OBJECT` for `Variant` type.

In `GDScriptCompiler::_gdtype_from_datatype` and `GDScriptCompiler::_parse_expression` trust analyzer with `builtin_type` of enum. (Irrelevant to this PR, just noticed, can remove.)

Closes #54312.
Closes #67368 (as superseded).

### CODE REVIEW

**Review by vnen — CHANGES_REQUESTED** [review]

File: modules/gdscript/gdscript_byte_codegen.cpp [hunk 1]
```diff
@@ -607,18 +607,45 @@ void GDScriptByteCodeGenerator::write_binary_operator(const Address &p_target, V
...
+				append(p_source);
+				append(p_type.builtin_type);
+			}
+		} break;
+		case GDScriptDataType::NATIVE: {
+			const Variant &native_class = GDScriptLanguage::get_singleton()->get_global_array()[GDScriptLanguage::get_singleton()->get_global_map()[p_type.native_type]];
+			append_opcode(GDScriptFunction::OPCODE_TYPE_TEST_NATIVE);
+			append(p_target);
+			append(p_source);
+			append(get_constant_pos(native_class) | (GDScriptFunction::ADDR_TYPE_CONSTANT << GDScriptFunction::ADDR_BITS));
```
[the remark this answers is not in the document]
Comment (vonagam) [reply modules/gdscript/gdscript_byte_codegen.cpp:633]: Hah, I actually wanted to ask you why in some places `GDScriptNativeClass` is passed and in some `StringName` is enough. Ok, will change it to `StringName`.

**Comment by vnen** [comment]
`is_type_of` sounds backwards to me (that is, I would expect the order of arguments to be inverted). Maybe we could use `is_instance_of`? While builtin types aren't technically "instances", I think the name is more fitting.

**Comment by vonagam** [comment]
I thought that in the future there might be a unified representation for a type (like `Variant::TYPE` or something) and with that in mind `is_type_of` would be more direct than `is_instance_of` (with second argument being of that type). But `is_instance_of` is obvious alternative, have no problem with that.

**Comment by vonagam** [comment]
Renamed to `is_instance_of`.

**Review by vnen — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: modules/gdscript/tests/scripts/analyzer/features/type_test_usage.gd [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	assert((b_array is Variant) == true)
+	assert((b_array is Array) == true)
+	assert(is_instance_of(b_array, TYPE_ARRAY) == true)
+	assert((b_array is Array[B]) == true)
+	assert((b_array is Array[A]) == false)
+	assert((b_array is Array[int]) == false)
+	assert((b_array is int) == false)
+	assert(is_instance_of(b_array, TYPE_INT) == false)
+
+	var const_b_array: Variant = [] as Array[B]
```
Comment (Iniquitatis) [inline modules/gdscript/tests/scripts/analyzer/features/type_test_usage.gd:51]:
```suggestion
	const const_b_array: Variant = [] as Array[B]
```

**Review by Iniquitatis — CHANGES_REQUESTED** [review]

**Pull request merged**